### ISSUE TITLE
Missing headers from livesim2 to receiver

### ISSUE THREAD

**Issue by nvkhoi112358** [Dash-Industry-Forum/livesim2#242]
livesim2 version 1.6.0

**Procedure**:
1. Start livesim2 and cmaf-ingest-receiver

```
./livesim2 --logformat json --loglevel debug --vodroot app/testdata/assets/
./cmaf-ingest-receiver -loglevel debug
```

2. Trigger pushing cmaf

- Non-low-latency

```
curl --request POST \
     --url http://localhost:8888/api/cmaf-ingests \
     --header 'Accept: application/json, application/problem+json'   --header 'Content-Type: application/json' \
     --data '{"destRoot": "http://localhost:8080/upload","destName": "testpic_2s","duration": 2000,"livesimURL": "/livesim2/testpic_2s/Manifest.mpd","streamsURLs": false}'
```

- Low latency

```
curl --request POST \
     --url http://localhost:8888/api/cmaf-ingests \
     --header 'Accept: application/json, application/problem+json'   --header 'Content-Type: application/json' \
     --data '{"destRoot": "http://localhost:8080/upload","destName": "testpic_2s","duration": 2000,"livesimURL": "/livesim2/ato_1.5/chunkdur_0.5/testpic_2s/Manifest.mpd","streamsURLs": false}'
```

**Expected**:
- Non-low-latency: `Content-Length` header is present at receiver
- Low latency: `Transfer-Encoding: chunked` is present at receiver

**Actual**:
- Non low latency: `Content-Length` is not there

```
time=2024-12-30T11:26:22.004+07:00 level=DEBUG msg=Headers chName=testpic_2s trName=V300 path=/testpic_2s/V300/867766390.cmfv headers="map[Accept-Encoding:[gzip] Connection:[keep-alive] Content-Type:[video/mp4] Dash-If-Ingest:[1.1] User-Agent:[Go-http-client/1.1]]"
...
2024/12/30 11:26:22 "PUT http://localhost:8080/upload/testpic_2s/V300/867766390.cmfv HTTP/1.1" from 127.0.0.1:54264 - 200 0B in 1.471734ms
```

- low latency: `Transfer-Encoding: chunked` is not there

```
time=2024-12-30T11:28:38.532+07:00 level=DEBUG msg=Headers chName=testpic_2s trName=A48 path=/testpic_2s/A48/867766459.cmfa headers="map[Accept-Encoding:[gzip] Connection:[keep-alive] Content-Type:[audio/mp4] Dash-If-Ingest:[1.1] User-Agent:[Go-http-client/1.1]]"
...
2024/12/30 11:28:40 "PUT http://localhost:8080/upload/testpic_2s/A48/867766459.cmfa HTTP/1.1" from 127.0.0.1:49898 - 200 0B in 1.502963168s

```

**What I found**:
- Direct stream from livesim2 still include those headers
- I tried to manual copy `cs.Header()` to `req.Header` right before https://github.com/Dash-Industry-Forum/livesim2/blob/14e9f6d28e133b062fb3c1dcca18021776e7bea0/cmd/livesim2/app/cmaf-ingester.go#L645 but nothing changes.

**Comment by tobbee** [comment]
@nvkhoi112358 Thanks for your report. 

These are headers that can be set by the HTTP client, but I don't think they are mandatory by the HTTP/1.1 standard and a receiver should not rely on them.

A Content-Length field would help the receiver to know how much total memory to allocate, but apart from that there is not a big difference compared to sending one big HTTP chunk.

I looked a bit at the Go HTTP client implementation, and I don't see an easy way to set the Content-Length in the HTTP Client. This is pretty logical since the API is an `io.Reader` interface which does not have the notion of a total size.

Regarding `Transfer-Encoding: chunked`, it is superfluous. If there is no Content-Length, the length of the content must be transferred using the chunk mechanism.

Why do you think you need these headers?

**Comment by nvkhoi112358** [comment]
Thank @tobbee 

> Do you see a big need for the Transfer-Encoding: chunked signalling?

Either `Content-Length` or `Transfer-Encoding` is good.

> I'll try to modify the CMAF ingester to set the Content-Length using a bytes.Buffer as indicated above.

But IMO, `Content-Length` is more important. Some conditions are never triggered without that header, and reading many of 1024 bytes is inefficient compares to reading all at one go.

**Comment by tobbee** [comment]
Status update: I'm working on this and got the Content-Length inserted properly, by reading all data into a bytes.Buffer if there is a Content-Length header set from the segment Writer. However, mixing this with the low-latency chunk-based approach is tricky, so I'm currently trying to understand how to synchronise the different Go-routines and writers and reader for that case...

**Comment by tobbee** [comment]
@nvkhoi112358 There is now a PR where Content-Length is used for normal mode of segments. Actually, the whole sending mechanism is much simplified in this case. Have a look at #244, and see if it provides what you want?

**Comment by nvkhoi112358** [comment]
Yes, that is exactly what I expect. Thank a lot. Case is resolved.
